### PR TITLE
SSO config for GitLab and Jenkins, plus guide on using SOPS encryption

### LINKED ISSUE #12: SSO self-registration disabled
Story:
As a DevOps team maintaining a deployment of the Software Factory package (a.k.a. DI2-ME) I want SSO self-registration to be disabled, so that I am able to control who has access.

Acceptance Criteria:
* Airgap compatibility is maintained
* ~~SSO self-registration is automatically disabled by default~~ Docs on how to disable self registration if desired
* ~~No manual configuration is required (CaC only)~~

### LINKED ISSUE #14: Auth requires 2FA
Story:
As a DevOps team maintaining a deployment of the Software Factory package (a.k.a. DI2-ME), I want all users of the software factory to automatically be required to use 2FA via TOTP token if they are logging in via username/password, so that I can achieve associated NIST compliance goals without needing to do any custom configuration.

Acceptance criteria:
* Maintains airgap compatibility
* ~~All new users are automatically required to set up 2FA TOTP on first login~~ Docs on how to configure 2FA
* ~~Does not require any manual configuration (CaC only)~~

### LINKED ISSUE #19: Jenkins auth via SSO
Outcome:
We want to update the Software Factory package (a.k.a. DI2-ME) so that Jenkins is using SSO

Acceptance criteria:
* Is able to run in an airgap
* A current user of the software factory is able to log into Jenkins and receive basic user level of access
* ~~Does not require any manual configuration (CaC only)~~

### LINKED ISSUE #28: Move secrets out of the ConfigMap and into a Secret
Currently all Big Bang configuration is in the one values.yaml file which gets added to the `common` ConfigMap in K8s. We should pull out any secrets and put them in the `common` Secret instead.

### LINKED ISSUE #46: GitLab as SSO provider
Per [ADR #1](https://github.com/defenseunicorns/zarf-package-software-factory/blob/main/doc/adr/0001-gitlab-as-sso-provider-but-without-cac-support.md) we want to configure GitLab as the SSO provider, so that GitLab accounts can be used to log into the other services in the software factory.

### PR DESCRIPTION
**Pull request by RothAndrew**
Documentation and configuration for configuring SSO for GitLab and Jenkins in the DI2-ME software factory, in a way that is as declarative as possible without spending a lot of time on custom automation.

Fixes #46
Fixes #12 
Fixes #14
Fixes #19
Fixes #28

### CODE REVIEW

**Comment by RothAndrew** [comment]
@jeff-mccoy / @YrrepNoj this is what I'm coming up with so far. I'm not wild about it, but can't think of a better way at the moment that doesn't explode the scope of the project by turning it into a Yeoman generator or something. Looking for feedback. before I spend the next several days hammering through all the details.

The difficulty comes in needing to stand up the environment, then go into GitLab and create an Application, which will generate a nondeterministic Application ID and Secret, which will need to be committed to the configuration repo, which will need to be brought back into Gitea using what I'm calling a "day2" zarf.yaml, so that Flux can pick it up and redeploy Jenkins with the SSO configured. If we just push a commit to Gitea it will be blown away by any further updates that are brought into the environment.

image

I think it is possible to completely automate this eventua

**Comment by david--wright** [comment]
I'm not wild about this method of doing it either. We could use the [gitlab api](https://docs.gitlab.com/ee/api/applications.html) to create the application and programmaticly apply that to the day2 config; but as you mentioned that might beyond the scope of what we want to do here.

**Comment by RothAndrew** [comment]
We could definitely programmatically generate the GitLab Applications, which would save manual clicks for sure. That's not the major issue I'm trying to figure out how to tackle right now. We have been saying for a little while now that the first round of this SSO stuff is going to be a lot more manual as an "MVP" of sorts before we spend extra time to make it declarative.

**Comment by RothAndrew** [comment]
Due to the way Jenkins is configured, doing the SSO configuration needs to be done as configuration as code, in the helm values file. Which means we need to get the Application ID and Secret and add it into the Jenkins helm values file, and trigger a re-sync in Flux. The easiest way to do that would be to just push a new commit to the repo in Gitea but as mentioned above, that commit would be blown away if we do any new updates of the config repo from outside the cluster, which will be required for day 2 ops

**Comment by RothAndrew** [comment]
I think this is ready to go. I'm gonna try to turn my brain off and follow the docs from start to finish to see if everything flows well.

**Comment by RothAndrew** [comment]
@david--wright would love your eyes on this as well as you have time.

File: manifests/setup.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+metadata:
+  name: zarf-package-software-factory
+  namespace: flux-system
+spec:
+  interval: 1m
+  url: http://zarf-gitea-http.zarf.svc.cluster.local:3000/zarf-git-user/mirror__github.com__RothAndrew__zarf-package-software-factory.git
+  secretRef:
+    name: zarf-git-secret
+  ref:
+    branch: feature/gitlab-sso-docs
```
Comment (RothAndrew) [inline manifests/setup.yaml:22]: This needs to be changed back to `main` before this merges

**Comment by RothAndrew** [comment]
After a brief convo with @jeff-mccoy and @YrrepNoj, I'm gonna go ahead and merge without a formal review and approval, just so things don't get stale since they are busy doing other things. I still welcome feedback from all sources and am happy to revisit anything that's here.

Next on the list is docs for manual SSO config of Jira. After Jira comes Confluence and Nexus.

Also high on the priority list is to disable the istio ingresses for the DevOps tools from Big Bang Core like Grafana, Kibana, etc, so that they can be accessible by cluster admins, but not exposed without SSO. That will be coming very soon.

Also also, high on the priority list is looking into the Istio config that @jeff-mccoy mentioned to figure out how to make this work while still using `bigbang.dev` as the domain.

**Pull request merged**